async-nats/src/jetstream/pull.rs

```diff
@@ -0,0 +1,192 @@
+
+use crate::{Error, StatusCode, Subscriber};
+
+use super::{
+    consumer::{Consumer, PullConsumerConfig},
+    Context, JetStreamMessage,
+};
+
+impl Consumer<PullConsumerConfig> {
+    pub async fn request_batch<I: Into<BatchOptions>>(&self, batch: I) -> Result<(), Error> {
+        let subject = format!(
+            "{}.CONSUMER.MSG.NEXT.{}.{}",
+            self.context.prefix, self.info.stream_name, self.info.name
+        );
+
+        let payload = serde_json::to_vec(&batch.into())?;
+
+        self.context
+            .client
+            .publish_with_reply(subject, self.inbox.clone(), payload.into())
+            .await?;
+        Ok(())
+    }
+
+    pub async fn fetch(&mut self, batch: usize) -> Result<BatchStream<'_>, Error> {
+        BatchStream::batch(
+            BatchOptions {
+                batch,
+                expires: None,
+                no_wait: true,
+                ..Default::default()
+            },
+            self,
+        )
+        .await
+    }
+
+    pub async fn batch(
+        &mut self,
+        batch: usize,
+        expires: Option<usize>,
+    ) -> Result<BatchStream<'_>, Error> {
+        BatchStream::batch(
+            BatchOptions {
+                batch,
+                expires,
+                no_wait: false,
+                idle_heartbeat: Duration::default(),
+                ..Default::default()
+            },
+            self,
+        )
+        .await
+    }
+
+    pub async fn process(&mut self, batch: usize) -> PullStream<'_> {
+        PullStream::new(self, batch)
+    }
+}
+
+pub struct BatchStream<'a> {
+    pending_messages: usize,
+    subscriber: &'a mut Subscriber,
+    context: Context,
+}
+
+impl<'a> BatchStream<'a> {
+    async fn batch(
+        batch: BatchOptions,
+        consumer: &'a mut Consumer<PullConsumerConfig>,
+    ) -> Result<BatchStream<'a>, Error> {
+        if consumer.subscription.is_none() {
+            consumer.subscription = Some(
+                consumer
+                    .context
+                    .client
+                    .subscribe(consumer.inbox.clone())
+                    .await?,
+            );
+        }
+        consumer.request_batch(batch).await?;
+
+        Ok(BatchStream {
+            pending_messages: batch.batch,
+            subscriber: consumer.subscription.as_mut().unwrap(),
+            context: consumer.context.clone(),
+        })
+    }
+}
+
+impl<'a> Stream for BatchStream<'a> {
+    type Item = Result<JetStreamMessage, Error>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        if self.pending_messages == 0 {
+            return std::task::Poll::Ready(None);
+        }
+        match self.subscriber.receiver.poll_recv(cx) {
+            std::task::Poll::Ready(message) => {
+                if let Some(message) = message {
+                    if message.status == Some(StatusCode::TIMEOUT) {
```

Comment:
Can simplify this a bit.

```suggestion
  match message.status.unwrap_or(StatusCode::OK) {
    StatusCode::TIMEOUT => {},
    StatusCode::IDLE_HEARTBEAT => {},
  }
```

Reply:
sure!